discovery/dnsdisc/builder.nim

```diff
@@ -0,0 +1,238 @@
+{.push raises: [Defect]}
+
+import
+  std/[sequtils, strformat, strutils, tables],
+  stew/[base64, base32, byteutils, results],
+  nimcrypto/[hash, keccak],
+  ./tree
+
+export tree
+
+## A collection of utilities for constructing a Merkle Tree
+## encoding a list of ENR and link entries.
+## The tree consists of DNS TXT records.
+## 
+## Discovery via DNS is based on https://eips.ethereum.org/EIPS/eip-1459
+## 
+## This implementation is based on the Go implementation of EIP-1459
+## at https://github.com/ethereum/go-ethereum/blob/master/p2p/dnsdisc
+
+const
+  HashAbbrevSize = 1 + (16*13)/8                 # Size of an encoded hash (plus comma)
+  MaxChildren = 370 div toInt(HashAbbrevSize)    # 13 children
+
+type
+  BuilderResult*[T] = Result[T, string]
+
+  Subtree* = object # A coherent section of a larger tree. Can contain branch and leaf nodes.
+    subtreeRoot*: SubtreeEntry
+    subtreeEntries*: seq[SubtreeEntry]
+
+####################
+# Helper functions #
+####################
+
+proc toTXTRecord*(rootEntry: RootEntry): BuilderResult[string] =
```

Comment:
How come BuilderResult is its own type? Not quite following

Reply:
It is to indicate that `Result`s in the builder all have `string` as an error type. In this case the return type is also a string.
See for example [similar implementation in eth2](https://github.com/status-im/nimbus-eth2/blob/44f652f704f14c67e648f2b14dee652d44b858c5/beacon_chain/spec/keystore.nim#L128) and the [docs for `Result`](https://github.com/status-im/nim-stew/blob/master/stew/results.nim#L25-L29)